netlogo-core/src/main/fileformat/NLogoXMLLoader.scala

```diff
@@ -0,0 +1,328 @@
+    (Version.is3D && extension == "nlogo3d") || (!Version.is3D && extension == "nlogo")
+
+  private def isCompatible(uri: URI): Boolean = {
+    val extension = AbstractModelLoader.getURIExtension(uri)
+    extension.isDefined && isCompatible(extension.get)
+  }
+
+  private def makeDimensions3D(dimensions: WorldDimensions, minPzcor: Int, maxPzcor: Int,
+                               wrappingAllowedInZ: Boolean): WorldDimensions = {
+    new WorldDimensions3D(dimensions.minPxcor, dimensions.maxPxcor, dimensions.minPycor, dimensions.maxPycor,
+                          minPzcor, maxPzcor, dimensions.patchSize, dimensions.wrappingAllowedInX,
+                          dimensions.wrappingAllowedInY, wrappingAllowedInZ)
+  }
+
+  def readModel(uri: URI): Try[Model] = {
+    val text =
+      if (uri.getScheme == "jar")
+        Source.fromInputStream(uri.toURL.openStream).mkString
+      else
+        Source.fromURI(uri).mkString
+    readModel(text, AbstractModelLoader.getURIExtension(uri).getOrElse(""))
+  }
+
+  def readModel(source: String, extension: String): Try[Model] = {
+
+    if (isCompatible(extension)) {
+
+      val reader = XMLInputFactory.newFactory.createXMLStreamReader(new StringReader(source))
+
+      while (reader.hasNext && reader.next != XMLStreamConstants.START_ELEMENT) {}
+
+      val element = readXMLElement(reader)
+
+      reader.close()
+
+      element.name match {
+        case "model" =>
+
+          import Model.{ defaultCode, defaultShapes, defaultLinkShapes }
+
+          val version = element("version")
+
+          val model = Model( defaultCode, List(), defaultInfo, version, defaultShapes, defaultLinkShapes, List(), Seq(), Seq())
+
+          element.children.foldLeft(Try(model)) {
+            case (model, XMLElement("widgets", _, _, children)) =>
+              model.map(_.copy(widgets = children.map(WidgetXMLLoader.readWidget(_, makeDimensions3D))))
+            case (model, XMLElement("info", _, text, _)) =>
+              model.map(_.copy(info = text))
+            case (model, XMLElement("code", _, text, _)) =>
+              model.map(_.copy(code = text))
+            case (model, el @ XMLElement("turtleShapes", _, _, _)) =>
+              model.map(_.copy(turtleShapes = el.getChildren("shape").map(ShapeXMLLoader.readShape)))
+            case (model, el @ XMLElement("linkShapes", _, _, _)) =>
+              model.map(_.copy(linkShapes = el.getChildren("shape").map(ShapeXMLLoader.readLinkShape)))
+            case (model, XMLElement("previewCommands", _, commands, _)) =>
+              val section = new Section("org.nlogo.modelsection.previewcommands", PreviewCommands.Custom(commands))
+              model.map((m) => m.copy(optionalSections = m.optionalSections :+ section))
+
+            // case "systemDynamics" =>
+            //   optionalSections = optionalSections :+
+            //     new Section("org.nlogo.modelsection.systemdynamics.gui", SDMXMLLoader.readDrawing(element))
```

Comment:
Is it intentional that this never got implemented?  If so: We should probably remove these lines.  If not: We should implement it.

Reply:
I thought we discussed this and arrived at the conclusion that the serialization/deserialization code could live in the SDM package, and then just be exposed through an API.  What came of that idea?

Reply:
I remember there being issues, but that was a long time ago. I'll revisit it.